test/src/com/legrig/CpsMismatchInConstructorSpec.groovy

```diff
@@ -0,0 +1,65 @@
+package com.legrig
+
+import com.cloudbees.groovy.cps.impl.CpsCallableInvocation
+import support.jenkins.CPSSpecification
+
+class CpsMismatchInConstructorSpec extends CPSSpecification {
+
+  Class testSubjectClass = CpsMismatchInConstructor
+
+  def "CpsMismatch(false) CPS Test"() {
+
+    when: 'we call constructor'
+      def inst = asCPSScript '''
+import com.legrig.CpsMismatchInConstructor
+
+new CpsMismatchInConstructor(false)
+'''
+    then: 'no exception is thrown'
+      noExceptionThrown()
+
+    and: 'inst is set'
+      inst instanceof CpsMismatchInConstructor
+  }
+
+  def "CpsMismatch(false) NonCPS Test"() {
+
+    when: 'we call constructor'
+      def inst = new CpsMismatchInConstructor(false)
+
+    then: 'no exception is thrown'
+      noExceptionThrown()
+
+    and: 'inst is set'
+      inst instanceof CpsMismatchInConstructor
+  }
+
+  def "CpsMismatch(true) CPS Test"() {
```

Comment:
cool! if this works correctly. Why does the following tests fail. I think there is some difference between how asCPSScript and invokeCPSMethod works. asCPSScript works as expected but invokeCPSMethod does not.  

when: invokeCPSMethod(sut, "nonCpsMethod")
then: 1==1

when: invokeCPSMethod(sut, "nonCpsMethodCallingCpsMethod")
then: thrown IllegalStateException;

Reply:
To be clear `asCPSScript()` just compiles the script from text with CPS enabled and then calls `invokeCPSMethod()` on it.

It makes sense that `invokeCPSMethod()` would not work on NonCPS code. It is not supposed to. The reason it works with `asCPSScript()` is because it is wrapped by CPS transformed script and CPS can call NonCPS. 

But invoking NonCPS code with `invokeCPSMethod()` directly will throw exception (`InvalidCPSInvocation`), as it is supposed to.

If you want to just run it as Jenkins runs it, the easiest thing to do is use `asCPSScript()` as it will guarantee you invoke CPS code no matter what.

Does this make sense?

Reply:
Hmm, I added some more tests and I think I found what you are talking about  - there is a corner case of "Mismatch Handling" when we are calling NonCPS code, thinking it is CPS code - but it calls CPS later - it does not fail, but also does not do the right thing.... This is not a scenario that can easily happen in Jenkins, and to do this, as a dev you really should be aware of what you are calling.  

I am not sure how to catch this, now that `isCPSTransformed()` is not working... For time being just use `asCPSScript()`

Reply:
yes, that makes sense. That is my understanding as well.

I see what you mean by "this is not a scenario that can easily happen in jenkins" as the entry flow definition is a cps transformed script. But the test driver asCPSScript/invokeCPSMethod should provide the same guarantees. asCPSScript provides that but invokeCPSMethod does not.

yes, asCPSScript does what I expect it to do. I brought this issue so that we track both asCPSScipt and invokeCPSMethod to behave similarly if that is the intention.

Reply:
may be we should apply CPSTransformer for the tests as well. That would make the tests to be CPS transformed and a non cps call from the test using invokeCPSMethod would work as expected.

Reply:
CPS Transforming the entire test directory is a non-starter - there is no way JUnit/Spock/etc will ever handle that safely - they are not written with CPS in mind.

The reason I say it is unlikely to happen on Jenkins is because the actual corner-case here is if you _START_ running NonCPS code while _thinking_ it is CPS and _THEN_ call a CPS method - there is no mismatch as it has no idea what to match it to - but in pipelines you always start in CPS mode, which means there you will always get a mismatch - so `asCPSScript` is the best approximation of that.

There is good news and not so good news, the good news is that `isCPSTransformed()` works just fine... until you wrap the object with GroovySpy or other mock/stub code.... Because it is wrapped, it cannot find the right annotations. without mock, it will work just fine (but then you are SOL on mocking/stubbing)

Right now the cost of breaking the GroovySpy wrappers is much higher than paying attention to own code you are testing, so I am not going to fix it. I will talk to Spock folks, maybe they know how to get around this issue

Reply:
CPSTransformer should just transform the files in test sourceset, 
JUnit/Spock are binaries, so those shouldn't be transformed.

Reply:
You can try it. Being that each of those classes extends and is transformed by Spock, I do not see it working, stranger things have happened :-)

Reply:
Got error at test compile time due to CPSTransformer. But error isn't
clear on what the issue is. No Spock classes referenced but Yes strange
things are happening.

`/CpsNonCPSUsageTest.groovy: -1: Unsupported expression for CPS transformation @ line -1, column -1.`